craft_providers/base.py

```diff
@@ -83,14 +266,627 @@ def get_command_environment(
+
+    def _post_setup_packages(self) -> None:
+        """Configure the new installed packages.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _pre_setup_snapd(self) -> None:
+        """Do anything before setting up snapd.
+
+        e.g.
+            enable depended services.
+            check possible incompatibility issues.
+
+        This step should be overridden when needed.
+        """
+        self._enable_udevd_service()
+        self._disable_snapd_cdn()
+
+    @abstractmethod
+    def _setup_snapd(self) -> None:
+        """Set up snapd.
+
+        This step must be overridden.
+        """
+
+    def _post_setup_snapd(self) -> None:
+        """Configure the new installed snapd.
+
+        This step usually does not need to be overridden.
+        """
+        self._enable_snapd_service()
+        self._disable_and_wait_for_snap_refresh()
+        self._setup_snapd_proxy()
+
+    def _pre_setup_snaps(self) -> None:
+        """Do anything before setting up the snaps.
+
+        e.g.
+            remove non-snap old packages.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _setup_snaps(self) -> None:
+        """Set up the snaps.
+
+        e.g.
+            install snaps.
+
+        This step should be overridden when needed.
+        """
+        self._install_snaps()
+
+    def _post_setup_snaps(self) -> None:
+        """Configure the new installed snaps.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _pre_clean_up(self) -> None:
+        """Do anything before cleaning up.
+
+        e.g.
+            stop affected services.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _clean_up(self) -> None:
+        """Cleanup the OS environment.
+
+        e.g.
+            remove unnecessary packages.
+            remove cache files.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _post_clean_up(self) -> None:
+        """Do anything needed after cleaning up.
+
+        e.g.
+            restart affected services.
+            rebuild packages database.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    def _pre_finish(self) -> None:
+        """Do anything needed before finishing the setup process.
+
+        e.g.
+            get the final package / snap versions.
+
+        This step should be overridden when needed.
+        """
+        return
+
+    @final
+    def _finish(self) -> None:
+        """Finish the setup process and update the status.
+
+        This step cannot be overridden.
+        """
+        self._update_setup_status(status=True)
+
+    @final
     def setup(
         self,
         *,
         executor: Executor,
-        retry_wait: float = 0.25,
-        timeout: Optional[float] = None,
+        timeout: Optional[float] = -1,
```

Comment:
Can we talk about the design for timeouts?  It can be positive, negative, 0, or None, but I don't see the behavior explained in the docstrings.  I'm asking because it looks more configurable than necessary.  We can place restrictions on the timeout parameter and simplify the code.

How about this implementation?  
Remove the `timeout` from `setup()` and `warmup()`.  Add a parameter to `__init__()` called `timeout: bool  = True`.  If true, the default timeouts are used.  If false, no timeouts are used.  I think that is enough.  You could even make the timeouts private (or make them class properties) to ensure they are always greater than 0.  This would simplify code in places like `_setup_wait_for_system_ready()` where you have to check if the timeout is greater than 0.

Reply:
I see.  I still have some concerns.  A negative timeout only affects the `unpredictable` timeout, one of 3 timeouts.  And it doesn't make a difference, it just sets it to the same value it was already initialized with.  A negative and 0 timeout have the same behavior and it isn't clear which should be used.

Reply:
That `-1` is just used to distinguish if this is the default value, since None is another legal value here. I think a timeout `0` may lead to some confusion, and the behaves differently from None.

Reply:
OK, that seems good. And use `None` to disable all timeout?

Reply:
Yes!